[LFI] Add MCLFIRewriter infrastructure

zyedidia opened the pull request:
This is the second patch in the LFI series, adding the following features:

* `MCLFIRewriter` class, which will be used to perform LFI rewrites on a per-architecture basis. Each architecture where LFI is supported will implement a subclass (for example, `Target/AArch64/MCTargetDesc/AArch64MCLFIRewriter.cpp`) that will implement the `rewriteInst` function to perform the actual rewriting (the AArch64 version will be added in the next PR). The generic rewriter class provides some instruction info utilities (`mayLoad`, `mayStore`) and is used to call `rewriteInst` during instruction emission. It also provides `onLabel` which allows the rewriter to know possible branch targets, making certain optimizations like guard elimination possible.
* LFI streamer initialization that marks object files with a NOTE section to indicate that the object file is using LFI.
* A basic LFI assembly parser that introduces the `.lfi_rewrite_disable` and `.lfi_rewrite_enable` directives that can be used to control whether rewriting is enabled or not in hand-written assembly.

zyedidia committed:
[LFI] Add MCLFIRewriter infrastructure

aengelke reviewed (commented):
Seems to be quite a bit of infrastructure that isn't yet testable..

zyedidia committed:
[LFI] Small fixes based on code review

- Use `unique_ptr` for `setLFIRewriter`.
- Fix `lfi-enable-rewriter` flag description.
- No need to explicitly initialize LFIRewriter to null.

zyedidia committed:
[LFI] initializeMCLFIStreamer in createMCObjectStreamer

- Still needs manual initialization for use with createAsmStreamer.

zyedidia committed:
[LFI] Rely on rewriteInst to track Enabled

zyedidia committed:
[LFI] Some additional minor cleanup

We could go further and remove the helpers in MCLFIRewriter.cpp entirely
and just use InstrDesc directly from inside each rewriter.

aengelke reviewed (approved):
LGTM with nits.

zyedidia committed:
Update comment headers based on CodingStandards

zyedidia committed:
Remove LLVM_ABI annotations

These will be added as part of a separate PR.

zyedidia committed:
Document .lfi_rewrite_enable/.lfi_rewrite_disable

aengelke reviewed (commented):
c-t impact on Clang seems larger than expected ([c-t-t](https://llvm-compile-time-tracker.com/compare_clang.php?from=d93ad10a2e9fb07132771cc5c9f356d4439c8950&to=f6c86bd69d92e82096fa98f8e38bb7ebf4ce9274&stat=instructions%3Au&sortBy=absolute-difference)). I think we could remove some includes.

nikic commented on llvm/include/llvm/MC/MCLFI.h:
```diff
@@ -0,0 +1,26 @@
+///
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Support/CommandLine.h"
```
This include looks unused?

zyedidia commented:
Okay thank you, sounds good. I have tried to reduce header usage as much as possible, and opened a PR: #182617. Is there any way we can test compile-time impact ahead of merging to confirm the improvement?

I'm also seeing these buildbot CI failures, but from what I can tell they seem unrelated to these changes. Let me know if there is any action to be taken on that front.

s-barannikov commented on llvm/lib/MC/MCLFI.cpp:
```diff
@@ -0,0 +1,78 @@
+  // Create the Target specific MCLFIRewriter.
+  assert(TheTarget != nullptr);
+  if (FlagEnableRewriting) {
+    TheTarget->createMCLFIRewriter(
```
Sounds good, thanks

zyedidia commented on llvm/tools/llvm-mc/llvm-mc.cpp:
```diff
@@ -626,6 +627,11 @@ int main(int argc, char **argv) {
 
+    Triple T(TripleName);
+    if (T.isLFI()) {
+      Str->initSections(NoExecStack, *STI);
```
I believe this was required since `initializeLFIMCStreamer` uses `pushSection`/`popSection` to insert the LFI ABI Note section, so there must be an existing section to avoid a crash. See this failure in the pending PR that performs rewrites: https://github.com/llvm/llvm-project/pull/186896#issuecomment-4107740880.

I'm happy to restructure this if you want a different approach -- please let me know if there is something that comes to mind.
